### PR TITLE
Get Kip500Broker into startable state

### PR DESCRIPTION
**Pull request by rondagostino**
Broker (almost) starts.  I see the broker sending a metadata request to identify the active controller, and it gets the response and correctly identifies that controller as the active one:

Here's the broker sending it:

```
[2020-11-08 18:25:59,972] INFO [broker-0-to-controller-send-thread]: Starting (kafka.server.Kip500BrokerToControllerRequestThread)
[2020-11-08 18:25:59,973] INFO [broker-0-to-controller-send-thread]: Active Controller isn't known, making metadata request to a random controller node to identify it (kafka.server.Kip500BrokerToControllerRequestThread)
[2020-11-08 18:25:59,974] INFO [broker-0-to-controller-send-thread]: Sending metadata request data MetadataRequestData(topics=[], allowAutoTopicCreation=false, includeClusterAuthorizedOperations=false, includeTopicAuthorizedOperations=false) to node localhost:9093 (id: 3000 rack: null) (kafka.server.Kip500BrokerToControllerRequestThread)
```

Here's the controller processing it (I locally added some logging for this):

```
[2020-11-08 18:26:00,035] INFO Received request: Request(processor=0, connectionId=127.0.0.1:9093-127.0.0.1:60813-0, session=Session(User:ANONYMOUS,/127.0.0.1), listenerName=ListenerName(PLAINTEXT), securityProtocol=PLAINTEXT, buffer=null) (kafka.server.ControllerApis)
[2020-11-08 18:26:00,036] INFO Handling {topics=[],allow_auto_topic_creation=false,include_cluster_authorized_operations=false,include_topic_authorized_operations=false,_tagged_fields={}} (kafka.server.ControllerApis)
```

And here's the broker receiving and processing it correctly:

```
[2020-11-08 18:26:00,048] INFO [broker-0-to-controller-send-thread]: Received metadata response org.apache.kafka.common.requests.MetadataResponse@8e446d5 (kafka.server.Kip500BrokerToControllerRequestThread)
[2020-11-08 18:26:00,050] INFO [broker-0-to-controller-send-thread]: Active controller is defined: localhost:9093 (id: 3000 rack: null) (kafka.server.Kip500BrokerToControllerRequestThread)
```

Then the broker heartbeat sends the registration request:

```
[2020-11-08 18:26:00,053] INFO [broker-0-to-controller-send-thread]: top request=BrokerToControllerQueueItem(org.apache.kafka.common.requests.BrokerRegistrationRequest$Builder@139347c9,kafka.server.BrokerLifecycleManagerImpl$$Lambda$325/548482954@41029111) (kafka.server.Kip500BrokerToControllerRequestThread)
```

And the controller processes it:

```
[2020-11-08 18:26:00,054] INFO Received request: Request(processor=0, connectionId=127.0.0.1:9093-127.0.0.1:60813-0, session=Session(User:ANONYMOUS,/127.0.0.1), listenerName=ListenerName(PLAINTEXT), securityProtocol=PLAINTEXT, buffer=null) (kafka.server.ControllerApis)
[2020-11-08 18:26:00,055] INFO Handling {broker_id=0,listeners=[{name=PLAINTEXT,host=localhost,port=9092,security_protocol=0,_tagged_fields={}}],features=[],rack=,_tagged_fields={}} (kafka.server.ControllerApis)
[2020-11-08 18:26:00,058] INFO Done handling {broker_id=0,listeners=[{name=PLAINTEXT,host=localhost,port=9092,security_protocol=0,_tagged_fields={}}],features=[],rack=,_tagged_fields={}} (kafka.server.ControllerApis)
```

But unfortunately the broker's request still times out.  I'm not sure why since the heartbeat request/response worked fine -- I would assume the registration request/response would also work fine.

```
[2020-11-08 18:26:18,073] ERROR [KafkaBroker id=0] Fatal error during controller startup. Prepare to shutdown (kafka.server.Kip500Broker)
java.util.concurrent.TimeoutException: Future timed out after [18000 milliseconds]
	at scala.concurrent.impl.Promise$DefaultPromise.tryAwait0(Promise.scala:212)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:225)
	at scala.concurrent.Await$.$anonfun$result$1(package.scala:201)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:62)
	at scala.concurrent.Await$.result(package.scala:124)
	at kafka.server.BrokerLifecycleManagerImpl.start(BrokerLifecycleManager.scala:179)
	at kafka.server.Kip500Broker.startup(Kip500Broker.scala:245)
	at kafka.server.KafkaServerManager.startup(KafkaServerManager.scala:110)
	at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:44)
	at kafka.Kafka$.main(Kafka.scala:82)
	at kafka.Kafka.main(Kafka.scala)
[2020-11-08 18:26:18,075] INFO [KafkaBroker id=0] shutting down (kafka.server.Kip500Broker)
```

### CODE REVIEW

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala [hunk 1]
```diff
@@ -93,7 +93,7 @@ class BrokerLifecycleManagerImpl(val brokerMetadataListener: BrokerMetadataListe
   private var schedulerTask: Option[ScheduledFuture[_]] = None
 
   // Broker states - Current and Target/Next
-  private var currentState: jmetadata.BrokerState = _
+  private var currentState: jmetadata.BrokerState = jmetadata.BrokerState.REGISTERING
```
Comment (aloknnikhil) [inline core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:96]: Thanks @rondagostino! I think this change would fail a couple of the tests in BrokerLifecycleManagerTest.scala considering they expect the initial state (of `NOT_RUNNING`) in some registration failure scenarios

**Review by aloknnikhil — DISMISSED** [review]

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala (hunk 1 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:96]: @aloknnikhil All tests pass, actually.  I learned today that recovery from an unclean shutdown happens synchronously when LogManager is instantiated -- the instantiation doesn't return until any recovery is complete.  So the heartbeat doesn't need to worry about that state at all, and it can jump right into registration.

Comment (aloknnikhil) [reply core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:96]: That's great. Definitely keeps it clean.

**Comment by rondagostino** [comment]
@aloknnikhil Looks like this may be a flaky test.  It failed and then succeeded upon retry.  Maybe it just needs a longer timeout since it took >900ms to succeed and the 1000ms timeout was the cause of the failure?

**Failure:**
```
[2020-11-30T22:42:27.756Z] kafka.server.KafkaConfigTest > testInterBrokerVersionMessageFormatCompatibility STARTED
[2020-11-30T22:42:27.756Z] kafka.server.metadata.BrokerMetadataListenerTest.testLocalLogManagerEventQueue failed, log available in /home/jenkins/workspace/entinc_Contributors_kafka_PR-447/core/build/reports/testOutput/kafka.server.metadata.BrokerMetadataListenerTest.testLocalLogManagerEventQueue.test.stdout
[2020-11-30T22:42:27.756Z] 
[2020-11-30T22:42:27.756Z] kafka.server.metadata.BrokerMetadataListenerTest > testLocalLogManagerEventQueue FAILED
[2020-11-30T22:42:27.756Z]     org.opentest4j.AssertionFailedError: Condition not met within timeout 1000. Wait for all events to be queued ==> expected: <true> but was: <false>
[2020-11-30T22:42:27.756Z]         at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
[2020-11-30T22:42:27.756Z]         at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:40)
[2020-11-30T22:42:27.756Z]         at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:193)
[2020-11-30T22:42:27.756Z]         at org.apache.kafka.test.TestUtils.lambda$waitForCondition$6(TestUtils.java:400)
[2020-11-30T22:42:27.756Z]         at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:448)
[2020-11-30T22:42:27.756Z]         at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:416)
[2020-11-30T22:42:27.756Z]         at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:397)
[2020-11-30T22:42:27.756Z]         at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:387)
[2020-11-30T22:42:27.756Z]         at kafka.server.metadata.BrokerMetadataListenerTest.testLocalLogManagerEventQueue(BrokerMetadataListenerTest.scala:290)
```

**Success:**
```
[2020-11-30T22:44:08.029Z] kafka.server.metadata.BrokerMetadataListenerTest > testLocalLogManagerEventQueue STARTED
[2020-11-30T22:44:08.954Z] 
[2020-11-30T22:44:08.954Z] kafka.server.metadata.BrokerMetadataListenerTest > testLocalLogManagerEventQueue PASSED
```

File: config/kip500-broker-node.properties [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+############################# Server Basics #############################
+
+# The role of this server. Setting this puts us in kip-500 mode
+process.roles=broker
+
+# The connect string for the controller quorum
+controller.connect=3000@localhost:9093
+
+# The id of the controller. This must be different than the broker id.
+#controller.id=3000
```
Comment (cmccabe) [inline config/kip500-broker-node.properties:27]: should get rid of controller.id and its comment, right?

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 3]
```diff
@@ -107,10 +107,34 @@ object Partition extends KafkaMetricsGroup {
   def apply(topicPartition: TopicPartition,
             time: Time,
             replicaManager: ReplicaManager): Partition = {
-    val zkIsrBackingStore = new ZkPartitionStateStore(
-      topicPartition,
-      replicaManager.zkClient,
-      replicaManager)
+    val backingStore: PartitionStateStore = if (replicaManager.zkClient.isDefined) {
```
Comment (cmccabe) [inline core/src/main/scala/kafka/cluster/Partition.scala:110]: It seems like implementing the interface, but then throwing `IllegalStateException` when we try to call most of the methods, is kind of ugly here.

What I think would be good here is to have three implementations of the interface: the existing ZkPartitionStateStore one, one that uses AlterIsr, and another one that uses either ZK or AlterIsr based on the IBP.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 4]
```diff
@@ -58,10 +58,20 @@ class GroupMetadataManager(brokerId: Int,
                            interBrokerProtocolVersion: ApiVersion,
                            config: OffsetConfig,
                            val replicaManager: ReplicaManager,
-                           zkClient: KafkaZkClient,
+                           groupMetadataTopicPartitionCountFunc: () => Int,
                            time: Time,
                            metrics: Metrics) extends Logging with KafkaMetricsGroup {
 
+  def this(brokerId: Int,
```
Comment (cmccabe) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:65]: Do we need to have this glue code?  I only count 3 places where we create a GroupMetadataManager.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 5]
```diff
@@ -1333,6 +1333,34 @@ object GroupCoordinator {
     apply(config, zkClient, replicaManager, heartbeatPurgatory, joinPurgatory, time, metrics)
   }
 
+  def apply(config: KafkaConfig,
```
Comment (cmccabe) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:1336]: This looks very similar to the previous `apply` function.  Can we express the one in terms of the other to avoid code duplication?

File: core/src/main/scala/kafka/coordinator/transaction/ProducerIdManager.scala [hunk 6]
```diff
@@ -70,7 +70,7 @@ case class ProducerIdBlock(brokerId: Int, blockStartId: Long, blockEndId: Long)
   }
 }
 
-class ProducerIdManager(val brokerId: Int, val zkClient: KafkaZkClient) extends Logging {
+class ProducerIdManager(val brokerId: Int, val zkClient: KafkaZkClient) extends ProducerIdMgr with Logging {
```
Comment (cmccabe) [inline core/src/main/scala/kafka/coordinator/transaction/ProducerIdManager.scala:73]: I realize this is just a first cut, but can we have the trait be named something like `ProducerIdGenerator`?  It's confusing to have both Manager and Mgr.
Also, it would be nice if the `ProducerIdGenerator` were defined in this file rather than elsewhere

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 3 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/cluster/Partition.scala:110]: This interface seems like kind of a grab-bag to me.  I think there's really 2 separate things going on here.  There's a `"TopicConfigRetriever"` (the `fetchTopicConfig(): Properties` method) and an `"IsrChangerViaZooKeeper"` (the `{shrink,expand}Isr(controllerEpoch: Int, leaderAndIsr: LeaderAndIsr): Option[Int]` methods).  The topic config retriever is used in **all** configurations (whenever a new Log has to be created), but the ZK Isr changer is **only** used when ZooKeeper is the AlterIsr communication mechanism.  What do you think about explicitly separating these two things out and having `Partition` take a mandatory topic config retriever and an `Option[]` for the ZK Isr changer?  When the partition instance gets constructed it can do the check on the ZK Isr changer Option to make sure it actually has something when `useAlterisr` is false.

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 7]
```diff
@@ -60,15 +60,15 @@ class LogManager(logDirs: Seq[File],
                  val retentionCheckMs: Long,
                  val maxPidExpirationMs: Int,
                  scheduler: Scheduler,
-                 val brokerState: AtomicReference[BrokerState],
+                 val cleanShutdownFunc: Boolean => Unit, // we send true if the shutdown was clean, otherwise false
```
Comment (cmccabe) [inline core/src/main/scala/kafka/log/LogManager.scala:63]: I think if you want to go this route you should make this a `CompletableFuture[Boolean]`, and remove `BrokerState` completely from `LogManager.scala`.

File: core/src/main/scala/kafka/server/AbstractBrokerToControllerChannelManager.scala [hunk 8]
```diff
@@ -38,28 +39,36 @@ trait BrokerToControllerChannelManager {
...
-class BrokerToControllerChannelManagerImpl(metadataCache: kafka.server.MetadataCache,
-                                           time: Time,
-                                           metrics: Metrics,
-                                           config: KafkaConfig,
-                                           threadNamePrefix: Option[String] = None) extends BrokerToControllerChannelManager with Logging {
-  private val requestQueue = new LinkedBlockingDeque[BrokerToControllerQueueItem]
-  private val logContext = new LogContext(s"[broker-${config.brokerId}-to-controller] ")
-  private val manualMetadataUpdater = new ManualMetadataUpdater()
-  private val requestThread = newRequestThread
+abstract class AbstractBrokerToControllerChannelManager(time: Time,
```
Comment (cmccabe) [inline core/src/main/scala/kafka/server/AbstractBrokerToControllerChannelManager.scala:56]: I think we need to rethink this part.  Deep inheritance hierarchies make the code difficult to understand.  Here we have the control flow bouncing between BrokerToControllerManager, AbstractBrokerToControllerManager, and Kip500BrokerToControllerManager, which makes the code hard to understand.

It seems like we should just look at the MetadataCache to see who the current controller is (if there is one).  Then we could use the same code for the pre and post-KIP-500 cases.

Another thing we need to do here is hook up `MetaLogManager` to the `MetadataCache`, so that when the leader changes, that change is reflected in the cache.  cc @hachikuji

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala [hunk 9]
```diff
@@ -47,7 +47,7 @@ import scala.concurrent.duration._
  */
 trait BrokerLifecycleManager {
 
-  // Initiate broker registration and start the heartbeat scheduler loop
+  // Start the heartbeat scheduler loop with broker registration as initial state
```
Comment (cmccabe) [inline core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:50]: broker registration is not the initial state, though.  as per the KIP, we start off in NOT_RUNNING

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala (hunk 1 above)
Comment (cmccabe) [inline core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:96]: the initial state should be NOT_RUNNING

Comment (cmccabe) [reply core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:96]: I think we should move the stuff about broker registration out of this class, and just handle heartbeats here.  That avoids the really awkward blocking in the `start` function.  We should also try to avoid keeping the current BrokerState here (we already store that in Kip500Broker, let's not duplicate things)

File: core/src/main/scala/kafka/server/KafkaBroker.scala [hunk 10]
```diff
@@ -121,6 +125,19 @@ trait KafkaBroker extends Logging with KafkaMetricsGroup {
   def currentState(): BrokerState
   def clusterId(): String
 
+  // methods required for legacy/kip500-agnostic dynamic configuration
+  def getConfig(): KafkaConfig
```
Comment (cmccabe) [inline core/src/main/scala/kafka/server/KafkaBroker.scala:129]: In Kafka we don't preface accessors with "get"

So this should be config(), authorizer(), etc. etc.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 11]
```diff
@@ -238,6 +245,9 @@ class KafkaApis(val requestChannel: RequestChannel,
     // stop serving data to clients for the topic being deleted
     val stopReplicaRequest = request.body[StopReplicaRequest]
     apisUtils.authorizeClusterOperation(request, CLUSTER_ACTION)
+    if (controller == null) {
```
Comment (cmccabe) [inline core/src/main/scala/kafka/server/KafkaApis.scala:248]: Can we rename "controller" to "legacyController" to make it more clear what's going on here?  We can do that in a separate PR if it's easier

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 12]
```diff
@@ -238,6 +245,9 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+      throw new ClusterAuthorizationException(s"Request $request should never happen in KIP-500 mode.")
```
Comment (cmccabe) [inline core/src/main/scala/kafka/server/KafkaApis.scala:249]: Probably makes more sense as UnsupportedVersionException

File: core/src/main/scala/kafka/server/KafkaBroker.scala (hunk 10 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/server/KafkaBroker.scala:129]: > In Kafka we don't preface accessors with "get"
So this should be config(), authorizer(), etc. etc.

I changed them all to have parameterless definitions (e.g. config, authorizer, etc. rather than config(), authorizer(), etc.) since that seems to be appropriate when the methods have no side-effects (which these do not), and it also means less change throughout (no need to add all those empty parens all over the place).

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 3 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/cluster/Partition.scala:110]: @mumrah will take a look at this in KC-1264.  I'll close this comment thread for now

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 7 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/log/LogManager.scala]: >  think if you want to go this route you should make this a CompletableFuture[Boolean], and remove BrokerState completely from LogManager.scala.

Thanks, that simplified things.

File: core/src/main/scala/kafka/server/AbstractBrokerToControllerChannelManager.scala (hunk 8 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/AbstractBrokerToControllerChannelManager.scala:56]: I filed KC-1265 to follow up on this.

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala (hunk 1 above)
Comment (aloknnikhil) [reply core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:96]: **Registration**:
- I think moving it out should be OK. But considering the `BrokerLifecycleManager` owns the lifecycle and any state transitions, I figured registering the broker would be the first step in starting the lifecycle manager.

**BrokerState**:
- The Kip500Broker function references the state maintained by the `BrokerLifecycleManager`. I think it makes sense to use that as the source of truth and leave it in the lifecycle manager considering it's the LifecycleManager that either approves or denies the state  transitions based on the response from the controller.

**Comment by cmccabe** [comment]
let's commit this and keep fixing forward.  I may make some more comments but this is a great start.  Thanks @rondagostino !
